ethcore/sync/src/block_sync.rs

```diff
@@ -465,9 +465,11 @@ impl BlockDownloader {
 	/// Checks if there are blocks fully downloaded that can be imported into the blockchain and does the import.
 	pub fn collect_blocks(&mut self, io: &mut SyncIo, allow_out_of_order: bool) -> Result<(), BlockDownloaderImportError> {
 		let mut bad = false;
+		let mut reset = false;
 		let mut imported = HashSet::new();
 		let blocks = self.blocks.drain();
 		let count = blocks.len();
+
 		for block_and_receipts in blocks {
```

Comment:
Might be worth to turn this loop into a closure and return an `enum Importer { Reset, Bad, Imported(HashSet) }` instead of setting the flags.

That said, refactoring can be left to a separate PR.